### PR TITLE
Make TTSCache safer

### PR DESCRIPTION
**Pull request by forslund**
## Description

Add `__contains__` method to TTSCache, The cache contains a SHA if the SHA is
known and all expected files exists on disk.

This is handles unexpected system events in a more consistent manner and
will still be fast for the case where a new sentence needs to be
synthesized.

## How to test
- Ask Mycroft to "say hello"
- remove all files in `/tmp/mycroft/cache/Mimic2/` (or the TTS you're using)
- ask Mycroft to "say hello" again
- verify that an audio is produced 

## Contributor license agreement signed?
CLA [ yes ]

### CODE REVIEW

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Failed ([Results](https://reports.mycroft.ai/core/PR-2914)). 
Mycroft logs are also available: [skills.log](https://reports.mycroft.ai/core/PR-2914/logs/skills.log), [audio.log](https://reports.mycroft.ai/core/PR-2914/logs/audio.log), [voice.log](https://reports.mycroft.ai/core/PR-2914/logs/voice.log), [bus.log](https://reports.mycroft.ai/core/PR-2914/logs/bus.log), [enclosure.log](https://reports.mycroft.ai/core/PR-2914/logs/enclosure.log)

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2914))

**Review by krisgesling — COMMENTED** [review]
Nice one!

I also verified that removing only the phoneme files works as expected.

Are there any improvements to the TestCache unit tests that we could be making?

**Comment by forslund** [comment]
One can always create more tests :) I'll try to add some

**Comment by forslund** [comment]
A couple of tests has been added!

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2914))

File: mycroft/tts/cache.py [hunk 1]
```diff
@@ -156,6 +162,16 @@ def __init__(self, tts_config, tts_name, audio_file_type):
...
 
+    def __contains__(self, sha):
+        """The cache contains a SHA if it knows of it and it exists on disk."""
+        if sha not in self.cached_sentences:
+            return False  # Doesn't know of it
+        else:
+            # Audio file must exist, phonemes are optional.
+            audio, phonemes = self.cached_sentences[sha]
+            return (audio.exists() and
+                    phonemes is None or phonemes.exists())
```
Comment (krisgesling) [inline mycroft/tts/cache.py:173]: The tests made me realise that this isn't quite right. I think we need parentheses around the 2nd line
eg
`A and (B or C)`
Otherwise it gets evaluated as 
`(A and B) or C`

eg
```
>>> False and True or True
True
```

File: test/unittests/tts/test_cache.py [hunk 2]
```diff
@@ -121,3 +121,50 @@ def test_curate_cache(self, curate_mock):
...
+        self.assertFalse('animal' in self.tts_cache)
+
+    def test_hash_exists_and_files_ok(self):
+        self.assertTrue('kermit' in self.tts_cache)
+
+    def test_only_audio_hash_exists_and_files_ok(self):
+        self.assertTrue('fozzie' in self.tts_cache)
+
+    def test_hash_exists_and_files_bad(self):
+        self.assertFalse('piggy' in self.tts_cache)
```
Comment (krisgesling) [inline test/unittests/tts/test_cache.py:170]: Gobo should be added here right?

With the Bool comparison fix above this seems to pass

**Review by krisgesling — COMMENTED** [review]
Yay for tests! 

Thanks for putting that together, I think we caught an edge case.

File: mycroft/tts/cache.py (hunk 1 above)
Comment (forslund) [reply mycroft/tts/cache.py:173]: Ah, quite right. will fix

**Comment by forslund** [comment]
Thanks for noticing that mistake, Added the "gobo" check and corrected the logic snafu

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2914))

**Review by krisgesling — APPROVED** [review]
